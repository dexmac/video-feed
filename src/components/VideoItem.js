import React, { Component } from "react";
import PropTypes from "prop-types";
import ReactPlayer from "react-player";
import VideoItemHeader from "./VideoItemHeader";
import { ERR_LOADING_VIDEO_PREFIX } from "../constants";

class VideoItem extends Component {
  state = {
    shouldShowVideo: true,
    errorString: ""
  };

  onVideoError = url => {
    this.setState({ shouldShowVideo: false, errorString: url });
  };

  render() {
    const item = this.props.item;

    let VideoPlayer;
    if (this.state.shouldShowVideo) {
      VideoPlayer = (
        <ReactPlayer
          className="react-player"
          url={item ? item.url : null}
          key={item ? item.videoId : null}
          width="100%"
          height="100%"
          onError={() => this.onVideoError(item ? item.url : null)}
          playing
          muted
          playsinline
          config={{ file: { forceVideo: true } }}
        />
      );
    } else {
      // On Video Load Error
      VideoPlayer = (
        <div className="video-load-error-text">
          {ERR_LOADING_VIDEO_PREFIX} : {this.state.errorString}
        </div>
      );
    }

    return (
      <div className="video-item">
        <VideoItemHeader
          title={item && item.title ? item.title : ""}
          views={item && item.views ? item.views : null}
        />
        <div className="player-wrapper">{VideoPlayer}</div>
      </div>
    );
  }
}

VideoItem.propTypes = {
  item: PropTypes.shape({
    url: PropTypes.string,
    source: PropTypes.string,
    videoId: PropTypes.string,
    title: PropTypes.string,
    views: PropTypes.number
  })
};
export default VideoItem;
