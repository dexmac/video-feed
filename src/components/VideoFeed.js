import React, { Component } from "react";
import PropTypes from "prop-types";
import VideoItem from "../components/VideoItem";

class VideoFeed extends Component {
  render() {
    let feedItems = this.props.items ? this.props.items : [];
    return (
      <div className="video-feed">
        {feedItems.map((item, i) => {
          return <VideoItem item={item} key={i} />;
        })};
      </div>
    );
  }
}

VideoFeed.propTypes = {
  items: PropTypes.array
};

export default VideoFeed;
