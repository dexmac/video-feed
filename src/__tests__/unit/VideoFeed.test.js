import React from "react";
import VideoFeed from "../../components/VideoFeed";
import VideoItem from "../../components/VideoItem";

describe("VideFeed", () => {
  let props;
  let mountedVideoFeed;
  let items;

  /**
   * Clearing props before each test
   */
  beforeEach(() => {
    mountedVideoFeed = undefined;
    items = [
      {
        title: "Big Buck Bunny",
        type: "video",
        source: "url",
        url:
          "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4",
        views: 8820
      }
    ];
  });

  /**
   * Shallow rendering vs. snapshot test
   */
  it("renders correctly && matches snapshot", () => {
    const tree = shallow(<VideoFeed />);
    expect(tree).toMatchSnapshot();
  });

  /**
   * Mounts the component using Enzyme if it is not mounted yet
   * @returns {object} - returns the mounted component
   */
  const videoFeed = items => {
    if (!mountedVideoFeed) {
      mountedVideoFeed = mount(<VideoFeed items={items} />);
    }
    return mountedVideoFeed;
  };

  describe("when no items are loaded in the feed", () => {
    it("should display an empty feed, i.e., feed length should be 0`", () => {
      const myFeed = videoFeed();
      expect(myFeed.find(VideoItem).length).toBe(0);
    });
  });

  it("should display a video item feed with 1 element and match the snapshot`", function() {
    let myFeedManager = videoFeed(items);
    expect(myFeedManager).toMatchSnapshot();
    expect(myFeedManager.find(VideoItem).length).toBe(1);
  });
});
