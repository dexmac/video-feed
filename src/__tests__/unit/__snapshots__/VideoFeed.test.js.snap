// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideFeed renders correctly && matches snapshot 1`] = `
<div
  className="video-feed"
>
  ;
</div>
`;

exports[`VideFeed should display a video item feed with 1 element and match the snapshot\` 1`] = `
<VideoFeed
  items={
    Array [
      Object {
        "source": "url",
        "title": "Big Buck Bunny",
        "type": "video",
        "url": "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4",
        "views": 8820,
      },
    ]
  }
>
  <div
    className="video-feed"
  >
    <VideoItem
      item={
        Object {
          "source": "url",
          "title": "Big Buck Bunny",
          "type": "video",
          "url": "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4",
          "views": 8820,
        }
      }
      key="0"
    >
      <div
        className="video-item"
      >
        <VideoItemHeader
          title="Big Buck Bunny"
          views={8820}
        >
          <div
            className="video-item-header"
          >
            <VideoItemTitle
              className="video-item-title"
              title="Big Buck Bunny"
            >
              <div
                className="video-item-title"
              >
                Big Buck Bunny
              </div>
            </VideoItemTitle>
            <VideoItemViewsNumber
              className="video-item-views-number"
              viewsNumber={8820}
            >
              <div
                className="video-item-views-number"
              >
                Views: 8820
              </div>
            </VideoItemViewsNumber>
          </div>
        </VideoItemHeader>
        <div
          className="player-wrapper"
        >
          <ReactPlayer
            className="react-player"
            config={
              Object {
                "file": Object {
                  "attributes": Object {
                    "autoPlay": "true",
                    "muted": "true",
                  },
                },
              }
            }
            controls={false}
            height="100%"
            loop={false}
            muted={true}
            onBuffer={[Function]}
            onDuration={[Function]}
            onEnded={[Function]}
            onError={[Function]}
            onPause={[Function]}
            onPlay={[Function]}
            onProgress={[Function]}
            onReady={[Function]}
            onSeek={[Function]}
            onStart={[Function]}
            playbackRate={1}
            playing={true}
            playsinline={true}
            progressInterval={1000}
            style={Object {}}
            url="http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4"
            volume={null}
            width="100%"
            wrapper="div"
          >
            <div
              className="react-player"
              style={
                Object {
                  "height": "100%",
                  "width": "100%",
                }
              }
            >
              <Player
                activePlayer={[Function]}
                className="react-player"
                config={
                  Object {
                    "dailymotion": Object {
                      "params": Object {
                        "api": 1,
                        "endscreen-enable": false,
                      },
                      "preload": false,
                    },
                    "facebook": Object {
                      "appId": "1309697205772819",
                    },
                    "file": Object {
                      "attributes": Object {
                        "autoPlay": "true",
                        "muted": "true",
                      },
                      "forceAudio": false,
                      "forceDASH": false,
                      "forceHLS": false,
                      "hlsOptions": Object {},
                      "tracks": Array [],
                    },
                    "mixcloud": Object {
                      "options": Object {
                        "hide_cover": 1,
                      },
                    },
                    "soundcloud": Object {
                      "options": Object {
                        "buying": false,
                        "download": false,
                        "liking": false,
                        "sharing": false,
                        "show_comments": false,
                        "show_playcount": false,
                        "visual": true,
                      },
                    },
                    "twitch": Object {
                      "options": Object {},
                    },
                    "vimeo": Object {
                      "playerOptions": Object {
                        "autopause": false,
                        "autoplay": false,
                        "byline": false,
                        "portrait": false,
                        "title": false,
                      },
                      "preload": false,
                    },
                    "wistia": Object {
                      "options": Object {},
                    },
                    "youtube": Object {
                      "playerVars": Object {
                        "autoplay": 0,
                        "iv_load_policy": 3,
                        "modestbranding": 1,
                        "playsinline": 1,
                        "rel": 0,
                        "showinfo": 0,
                      },
                      "preload": false,
                    },
                  }
                }
                controls={false}
                height="100%"
                key="FilePlayer"
                loop={false}
                muted={true}
                onBuffer={[Function]}
                onDuration={[Function]}
                onEnded={[Function]}
                onError={[Function]}
                onPause={[Function]}
                onPlay={[Function]}
                onProgress={[Function]}
                onReady={[Function]}
                onSeek={[Function]}
                onStart={[Function]}
                playbackRate={1}
                playing={true}
                playsinline={true}
                progressInterval={1000}
                style={Object {}}
                url="http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4"
                volume={null}
                width="100%"
                wrapper="div"
              >
                <FilePlayer
                  activePlayer={[Function]}
                  className="react-player"
                  config={
                    Object {
                      "dailymotion": Object {
                        "params": Object {
                          "api": 1,
                          "endscreen-enable": false,
                        },
                        "preload": false,
                      },
                      "facebook": Object {
                        "appId": "1309697205772819",
                      },
                      "file": Object {
                        "attributes": Object {
                          "autoPlay": "true",
                          "muted": "true",
                        },
                        "forceAudio": false,
                        "forceDASH": false,
                        "forceHLS": false,
                        "hlsOptions": Object {},
                        "tracks": Array [],
                      },
                      "mixcloud": Object {
                        "options": Object {
                          "hide_cover": 1,
                        },
                      },
                      "soundcloud": Object {
                        "options": Object {
                          "buying": false,
                          "download": false,
                          "liking": false,
                          "sharing": false,
                          "show_comments": false,
                          "show_playcount": false,
                          "visual": true,
                        },
                      },
                      "twitch": Object {
                        "options": Object {},
                      },
                      "vimeo": Object {
                        "playerOptions": Object {
                          "autopause": false,
                          "autoplay": false,
                          "byline": false,
                          "portrait": false,
                          "title": false,
                        },
                        "preload": false,
                      },
                      "wistia": Object {
                        "options": Object {},
                      },
                      "youtube": Object {
                        "playerVars": Object {
                          "autoplay": 0,
                          "iv_load_policy": 3,
                          "modestbranding": 1,
                          "playsinline": 1,
                          "rel": 0,
                          "showinfo": 0,
                        },
                        "preload": false,
                      },
                    }
                  }
                  controls={false}
                  height="100%"
                  loop={false}
                  muted={true}
                  onBuffer={[Function]}
                  onDuration={[Function]}
                  onEnded={[Function]}
                  onError={[Function]}
                  onPause={[Function]}
                  onPlay={[Function]}
                  onProgress={[Function]}
                  onReady={[Function]}
                  onSeek={[Function]}
                  onStart={[Function]}
                  playbackRate={1}
                  playing={true}
                  playsinline={true}
                  progressInterval={1000}
                  style={Object {}}
                  url="http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4"
                  volume={null}
                  width="100%"
                  wrapper="div"
                >
                  <video
                    autoPlay="true"
                    controls={false}
                    loop={false}
                    muted="true"
                    preload="auto"
                    src="http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4"
                    style={
                      Object {
                        "height": "100%",
                        "width": "100%",
                      }
                    }
                  />
                </FilePlayer>
              </Player>
            </div>
          </ReactPlayer>
        </div>
      </div>
    </VideoItem>
    ;
  </div>
</VideoFeed>
`;
